src/Features/LanguageServer/Protocol/RoslynLanguageServer.cs

```diff
@@ -107,5 +110,71 @@ public Task OnInitializedAsync(ClientCapabilities clientCapabilities, RequestCon
             OnInitialized();
             return Task.CompletedTask;
         }
+
+        protected override string GetLanguageForRequest(string methodName, JToken? parameters)
+        {
+            if (parameters == null)
+            {
+                _logger.LogInformation("No request parameters given, using default language handler");
+                return LanguageServerConstants.DefaultLanguageName;
+            }
+
+            // For certain requests like text syncing we'll always use the default language handler
+            // as we do not want languages to be able to override them.
+            if (ShouldUseDefaultLanguage(methodName))
+            {
+                return LanguageServerConstants.DefaultLanguageName;
+            }
+
+            var lspWorkspaceManager = GetLspServices().GetRequiredService<LspWorkspaceManager>();
+
+            // All general LSP spec document params have the following json structure
+            // { "textDocument": { "uri": "<uri>" ... } ... }
+            //
+            // We can easily identify the URI for the request by looking for this structure
+            var textDocumentToken = parameters["textDocument"];
```

Comment:
@dibarbet I'm looking at supporting OnAutoInsert. At the minimum we will also need this:
``` csharp
var textDocumentToken = parameters["textDocument"] ?? parameters["_vs_textDocument"];
```

Reply:
good catch, absolutely needed.